gax-httpjson/src/main/java/com/google/api/gax/httpjson/HttpJsonStatusCode.java

```diff
@@ -31,25 +31,125 @@
 
 import com.google.api.core.BetaApi;
 import com.google.api.gax.rpc.StatusCode;
+import com.google.common.base.Strings;
 import java.util.Objects;
 
 /** A failure code specific to an HTTP call. */
 @BetaApi
 public class HttpJsonStatusCode implements StatusCode {
-  private final int code;
+  static final String FAILED_PRECONDITION = "FAILED_PRECONDITION";
+  static final String OUT_OF_RANGE = "OUT_OF_RANGE";
+  static final String ALREADY_EXISTS = "ALREADY_EXISTS";
+  static final String DATA_LOSS = "DATA_LOSS";
+  static final String UNKNOWN = "UNKNOWN";
 
-  /** Returns the status code from the http call. */
-  public int getCode() {
-    return code;
-  }
+  private final int httpStatus;
+  private final StatusCode.Code statusCode;
 
   /** Creates a new instance with the given status code. */
-  public static HttpJsonStatusCode of(int code) {
-    return new HttpJsonStatusCode(code);
+  public static HttpJsonStatusCode of(int httpStatus, String errorMessage) {
+    return new HttpJsonStatusCode(httpStatus, httpStatusToStatusCode(httpStatus, errorMessage));
+  }
+
+  public static HttpJsonStatusCode of(StatusCode.Code statusCode) {
+    return new HttpJsonStatusCode(statusCodeToHttpStatus(statusCode), statusCode);
+  }
+
+  private static StatusCode.Code httpStatusToStatusCode(int httpStatus, String errorMessage) {
```

Comment:
Maybe make this a `switch()` (seems like a proper job for it). Switches can have nested switches too (or just nested if-elses)?

Reply:
Will do.